### PR TITLE
Use an enum rather than JSON value for storing source params

### PR DESCRIPTION
**Pull request by guilload**
### Description
Use an enum rather than JSON value for storing source params, which indirectly validates the content of the params provided by the user. To get this PR out in a few hours for the release, I took a few shortcuts :/

### How was this PR tested?
In progress

### CODE REVIEW

File: quickwit-config/src/index_config.rs [hunk 1]
```diff
@@ -275,6 +277,18 @@ impl IndexConfig {
             bail!("Index config contains duplicate sources.")
         }
 
+        // We want to forbid file source with no filepath.
```
Comment (fulmicoton) [inline quickwit-config/src/index_config.rs:280]:
```suggestion
Refactor
```

Comment (fulmicoton) [inline quickwit-config/src/index_config.rs:280]:
```suggestion
        // TODO refactor. This should be a method on `SourceType`.
        // We want to forbid file source with no filepath.
```

File: quickwit-config/src/source_config.rs [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+use quickwit_common::uri::Uri;
+use serde::de::Error;
+use serde::{Deserialize, Deserializer, Serialize};
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct SourceConfig {
+    pub source_id: String,
+    #[serde(flatten)]
+    pub source_type: SourceType,
```
Comment (fulmicoton) [inline quickwit-config/src/source_config.rs:30]: I expect `SourceType` to be a `#[derive(Clone,Copy)] enum`.

Not something that holds a param object.
Maybe rename `SourceParams` ?

File: quickwit-config/src/source_config.rs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+pub struct FileSourceParams {
+    /// Path of the file to read. Assume stdin if None.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    #[serde(deserialize_with = "absolute_filepath_from_str")]
+    pub filepath: Option<PathBuf>, //< If None read from stdin.
+}
+
+// Deserializing a filepath string into an absolute filepath.
+fn absolute_filepath_from_str<'de, D>(deserializer: D) -> Result<Option<PathBuf>, D::Error>
```
Comment (fulmicoton) [inline quickwit-config/src/source_config.rs:98]: here it is canonical.
absolute is slightly different from canonical.
canonical resolve links.

It's raises an interesting question. Here we might actually prefer absolute.

File: quickwit-config/src/source_config.rs (hunk 2 above)
Comment (fmassot) [reply quickwit-config/src/source_config.rs:30]: yes indeed

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**